### PR TITLE
chore(tests): Add BuildPulse reporting

### PR DESCRIPTION
**Pull request by czeslavo**
**What this PR does / why we need it**:

Adding BuildPulse reporting to `test.yaml`, `e2e.yaml` and `nightly.yaml` workflows to enable flaky test detection. It uses [gotestsum tool](https://github.com/gotestyourself/gotestsum) to generate JUnit reports and `Workshop64/buildpulse-action` to send reports to BuildPulse.

`gotestsum` is used in every make target that was calling `go test`, and is replacing these calls with `gotestsum --`. In CI we define `GOTESTSUM_JUNITFILE` which is picked up by `gotestsum` and JUnit xml report is generated. 

In `*.pretty` make targets `GOTESTSUM_FORMAT` is defined to change the output format to a more concise and human-friendly one. 

**Which issue this PR fixes**:

Solves https://github.com/Kong/kubernetes-ingress-controller/issues/2985

**Special notes for your reviewer**:

I have removed `gotestfmt` in favour of `gotestsum` in `*.pretty` make targets as they both kinda overlap with the functionality of prettifying the output, while `gotestfmt` lacks JUnit reporting which is required for BuildPulse. 

```
make test.unit.pretty
✓  internal/controllers/utils (3.007s) (coverage: 1.5% of statements in ./pkg/..., ./internal/...)
✓  internal/annotations (3.008s) (coverage: 2.0% of statements in ./pkg/..., ./internal/...)
✓  internal/dataplane/kongstate (3.298s) (coverage: 9.8% of statements in ./pkg/..., ./internal/...)
✓  internal/dataplane/deckgen (3.264s) (coverage: 1.7% of statements in ./pkg/..., ./internal/...)
✓  internal/cmd/rootcmd (3.553s) (coverage: 1.4% of statements in ./pkg/..., ./internal/...)
✓  internal/admission (3.937s) (coverage: 3.2% of statements in ./pkg/..., ./internal/...)
✓  internal/controllers/gateway (4.759s) (coverage: 2.6% of statements in ./pkg/..., ./internal/...)
✓  internal/dataplane/parser (2.66s) (coverage: 19.7% of statements in ./pkg/..., ./internal/...)
✓  internal/meshdetect (2.184s) (coverage: 3.1% of statements in ./pkg/..., ./internal/...)
✓  internal/dataplane/parser/translators (1.497s) (coverage: 2.4% of statements in ./pkg/..., ./internal/...)
✓  internal/dataplane/sendconfig (1.759s) (coverage: 1.9% of statements in ./pkg/..., ./internal/...)
✓  internal/manager (1.968s) (coverage: 1.4% of statements in ./pkg/..., ./internal/...)
✓  internal/store (2.031s) (coverage: 5.3% of statements in ./pkg/..., ./internal/...)
✓  internal/util/kubernetes/object (1.601s) (coverage: 1.5% of statements in ./pkg/..., ./internal/...)
✓  internal/util/kubernetes/object/status (1.547s) (coverage: 1.4% of statements in ./pkg/..., ./internal/...)
✓  internal/validation/consumers/credentials (1.516s) (coverage: 1.5% of statements in ./pkg/..., ./internal/...)
✓  internal/adminapi (11.534s) (coverage: 1.9% of statements in ./pkg/..., ./internal/...)
∅  internal/cmd
∅  internal/controllers/configuration
∅  internal/controllers/knative
✓  internal/util (3.799s) (coverage: 4.2% of statements in ./pkg/..., ./internal/...)
✓  internal/validation/gateway (1.215s) (coverage: 2.0% of statements in ./pkg/..., ./internal/...)
✓  internal/dataplane (21.334s) (coverage: 2.2% of statements in ./pkg/..., ./internal/...)
∅  internal/diagnostics
∅  internal/manager/metadata
∅  internal/manager/utils
∅  internal/metrics
∅  internal/types
∅  internal/util/test
∅  internal/versions
∅  pkg/apis/configuration/v1
∅  pkg/apis/configuration/v1alpha1
∅  pkg/apis/configuration/v1beta1
∅  pkg/clientset
∅  pkg/clientset/fake
∅  pkg/clientset/scheme
∅  pkg/clientset/typed/configuration/v1
∅  pkg/clientset/typed/configuration/v1/fake
∅  pkg/clientset/typed/configuration/v1alpha1
∅  pkg/clientset/typed/configuration/v1alpha1/fake
∅  pkg/clientset/typed/configuration/v1beta1
∅  pkg/clientset/typed/configuration/v1beta1/fake

DONE 692 tests in 29.952s
```

Test results are going to be visible on [this page](https://buildpulse.io/@Kong/kubernetes-ingress-controller). 

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- ~~[ ] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR~~

### CODE REVIEW

File: .github/workflows/e2e.yaml [hunk 1]
```diff
@@ -115,3 +123,35 @@ jobs:
...
+
+      - name: Upload test results to BuildPulse for flaky test detection
+        if: ${{ !cancelled() }}
+        uses: Workshop64/buildpulse-action@main
+        with:
+          account: 962416
+          repository: 127765544
+          path: report/*.xml
+          key: ${{ secrets.BUILDPULSE_ACCESS_KEY_ID }}
+          secret: ${{ secrets.BUILDPULSE_SECRET_ACCESS_KEY }}
```
Comment (pmalek) [inline .github/workflows/e2e.yaml:157]: Be sure to add those to 1pass

File: Makefile [hunk 2]
```diff
@@ -211,7 +211,7 @@ KIND_CLUSTER_NAME ?= "integration-tests"
 INTEGRATION_TEST_TIMEOUT ?= "45m"
 E2E_TEST_TIMEOUT ?= "45m"
 KONG_CONTROLLER_FEATURE_GATES ?= GatewayAlpha=true
-GOTESTFMT_CMD ?= $(GOTESTFMT) -hide successful-downloads,empty-packages -showteststatus
+export GOTESTSUM_FORMAT=standard-verbose
```
Comment (pmalek) [inline Makefile:214]: What's the difference between a regular makefile assignment and this?

File: .github/workflows/e2e.yaml (hunk 1 above)
Comment (czeslavo) [reply .github/workflows/e2e.yaml:157]: Do you think it's gonna be needed even if the secrets are accessible for every team member via https://buildpulse.io/@Kong/kubernetes-ingress-controller/setup/github-actions?

File: Makefile (hunk 2 above)
Comment (czeslavo) [reply Makefile]: Without the `export`, we have to inject the env variable explicitly to every command, like the following: [839f341](https://github.com/Kong/kubernetes-ingress-controller/pull/3021/commits/839f341a75e7cac7c5a005b01fec1ef2ac8de7ab)

I think that will be less confusing 👍

**Review by pmalek — APPROVED** [review]
Nice 👍

File: .github/workflows/e2e.yaml (hunk 1 above)
Comment (pmalek) [reply .github/workflows/e2e.yaml:157]: We'd have to document that somewhere so that folks know how to find it

Comment (czeslavo) [reply .github/workflows/e2e.yaml:157]: https://github.com/Kong/team-k8s/pull/253

**Review by pmalek — APPROVED** [review]

**Pull request merged**